### ISSUE TITLE
Batch generation does not work correctly for networks without recurrency

### ISSUE THREAD

**Issue by JackTemaki** [rwth-i6/returnn#362]
RETURNN has a simple test for pushing data into hdf file (irrelevant parts are skipped):

```python
def test_engine_forward_to_hdf():
  [...]
  seq_len = 5
  n_data_dim = 2
  n_classes_dim = 3
  num_seqs = 20
  dataset = DummyDataset(input_dim=n_data_dim, output_dim=n_classes_dim,
                         num_seqs=num_seqs, seq_len=seq_len)
  [...]
  config.update({
    "model": "%s/model" % _get_tmp_dir(),
    "num_outputs": n_classes_dim,
    "num_inputs": n_data_dim,
    "network": {"output": {"class": "softmax", "loss": "ce"}},
    "output_file": output_file,
  })
  [...]
  engine.forward_to_hdf(data=dataset, output_file=output_file, batch_size=5)
  [...]
  with h5py.File(output_file, 'r') as f:
    [...]
    assert f['seqLengths'].shape == (num_seqs,2)
  [...]
```

This test would fail if the `batch_size` is not exactly the same as each `seq_len`. This also means that this would not work for any dataset that produces sequences of varying length.

The reason is, that the batch construction works in a different way if the network does not contain recurrency (from `Dataset._generate_batches()`:

```python
      else:  # Not recurrent.
        while t_start.max_value() < t_end.max_value():
          length = t_end - t_start
          num_frames = NumbersDict.min(
            [length, batch_size.copy_like(length) - batch.get_all_slices_num_frames().copy_like(length)])
          assert num_frames.max_value() > 0
          batch.add_frames(seq_idx=seq_idx, seq_start_frame=t_start, length=num_frames)
          if (batch.get_all_slices_num_frames().any_compare(batch_size, (lambda a, b: a >= b))
                  or batch.get_num_seqs() > max_seqs):
            yield batch
            batch = Batch()
          t_start += num_frames
```
This code puts exactly as many time frames in a batch as there are sequences and drops the time axis, meaning that sequences might be split in half. The `hdf_to_forward` code sees the output as single sequence, and will write the data itself correctly, but the `seqLength` entries wrong.

I am not yet sure how to fix this myself, so I am raising this issue.

**Comment by albertz** [comment]
The batch generation is fine and as intended. This is a feature.
The definition of `recurrent` (in a layer/network/loss in RETURNN) is that sequences and frame-order does not matter at all, so this is a valid optimization.

The problem is just that `forward_to_hdf` cannot handle it, and also we do not really save the information to recover the sequences.

One simple solution would be to do sth like this in `forward_to_hdf`:
```
    batches = data.generate_batches(
      recurrent_net=True,
...
```

This should already fix it.

**Comment by JackTemaki** [comment]
> The problem is just that `forward_to_hdf` cannot handle it, and also we do not really save the information to recover the sequences.

So it means that this is only intended to be used for training? 

> This should already fix it.

Yes ok, of course we can avoid using the non-recurrent batch construction code.

**Comment by albertz** [comment]
> > The problem is just that `forward_to_hdf` cannot handle it, and also we do not really save the information to recover the sequences.
> 
> So it means that this is only intended to be used for training?

There was no intention. Whoever wrote that code probably just did not think about this case.

In any case, we can do it just as we want.

> > This should already fix it.
> 
> Yes ok, of course we can avoid using the non-recurrent batch construction code.

I think this is by far the simplest solution.

**Comment by JackTemaki** [comment]
Closed with 18a54de3af58f50ceb6978afce29468dc98b47db.
